core/tests/protractor_utils/users.js

```diff
@@ -61,24 +61,66 @@ var login = async function(email, useManualNavigation = true) {
     await browser.get(general.SERVER_URL_PREFIX + general.LOGIN_URL_SUFFIX);
   }
 
+
   var loginPage = element(by.css('.protractor-test-login-page'));
   await waitFor.presenceOf(loginPage, 'Login page did not load');
 
   var emailInput = element(by.css('.protractor-test-sign-in-email-input'));
   await action.sendKeys('Email input', emailInput, email);
 
   var signInButton = element(by.css('.protractor-test-sign-in-button'));
+
+  // Login page performs client side redirection which is known to cause
+  // "both angularJS testability and angular testability are undefined" flake.
+  // As suggested protractor devs here (http://git.io/v4gXM), waiting for
+  // angular is disabled during client side redirects.
+  await browser.waitForAngularEnabled(false);
   await action.click('Sign in button', signInButton);
 
+  // The await action.click only waits until button is clicked, it doesnot wait
+  // for the redirection to trigger. So, manually waiting for redirection here.
+  await browser.driver.wait(async() => {
+    const URL = await browser.driver.getCurrentUrl();
+    // As redirect url can be any url, so instead of testing for new url,
+    // checking whether the url is not /login.
+    return !(/login/.test(URL));
```

Comment:
Can you please explain this line . I am a bit confused.

Reply:
The /login, /logout, /signup URL has a query parameter redirect_url, based on this parameter it redirects users after login/logout/signup is successful.
So, I cannot wait for the URL to be changed to the new URL (as it can be any URL), I'm waiting for the URL no longer be the present URL (this indicates the redirection event).